### PR TITLE
CODINGSTYLE: add object types section

### PR DESCRIPTION
**Pull request by morantr**
Add a standard for objects and types definition and usage, to try and minimize
the number of review comments related to it.

### CODE REVIEW

File: CODINGSTYLE.md
Comment (arnout) [inline CODINGSTYLE.md:28]: We discussed this a bit more than half a year ago. The code currently has roughly 50% sCamelCase and 50% snake_case. Back then, we decided that we would use snake_case (except to keep consistency within the same file if it is already using sCamelCase). Is there any reason to override that decision now?

File: CODINGSTYLE.md
Comment (arnout) [inline CODINGSTYLE.md:24]:
```suggestion
Structs shall only be used for "plain-old-data" POD type with no logic or any functions except a data member initialization function (constructor).
Structs shall not have protected or private members.
Structs shall not inherit, not shall others inherit from it.
```

File: CODINGSTYLE.md
Comment (arnout) [inline CODINGSTYLE.md:43]: Same here: we decided before to use lower snake case.

File: CODINGSTYLE.md
Comment (arnout) [inline CODINGSTYLE.md:63]: For this, no decision had been taken before. However, we currently have about 40 typedefs (not counting the 'typedef struct' cases which shouldn't be typedefs to begin with), and only 14 with `_t`. That said, looking at the ones without `_t`, they are either just shorthand for class names in other namespaces, or they are already violating the naming convention in a thousand other ways (I'm looking at you, bml...). So I think this is definitely a good rule.

File: CODINGSTYLE.md
Comment (arnout) [inline CODINGSTYLE.md]: The text below is really still about naming convention. Even "when to use struct and when to use class" can be considered naming convention. So I wouldn't add this subtitle and just make the rest subsections of Naming convention.

**Review by arnout — COMMENTED** [review]
I can't approve this PR, since it's actually a change in coding style / naming convention that was not discussed before. I'm personally a fan of the naming convention we had, because it's simple. And I was the one who wrote it down to begin with :-).

So please open a vote on #democracy before merging this PR.

File: CODINGSTYLE.md
Comment (morantr) [reply CODINGSTYLE.md:28]: The main reason I did it now is since you gave a comment about it on another PR of mine. 😄 
But besides that, I think that it's missing and we need a convention for that.
Using the same convention for type and variables name is annoying IMO since usually we would want to give the same name of the type to the variable itself.

Bad IMO:
```cpp
class some_class {
    // code
};

// This is annoying since it forces me to be creative and give the variable a different name from the type
some_class some_class_obj;
```

Good IMO:
```cpp
class SomeClass {
    // code
};

// Perfect IMO :)
SomeClass some_class;
```

File: CODINGSTYLE.md
Comment (morantr) [reply CODINGSTYLE.md:24]: good idea.

File: CODINGSTYLE.md
Comment (morantr) [reply CODINGSTYLE.md]: Ok, Removed the title.

File: CODINGSTYLE.md
Comment (morantr) [reply CODINGSTYLE.md:63]: thanks

File: CODINGSTYLE.md
Comment (tomereli) [reply CODINGSTYLE.md:28]: I tend to agree with @morantr on this one now, at least new classes should use camle case for the class definition, but NOT for struct definition which is what @arnout commented on here - https://github.com/prplfoundation/prplMesh/pull/1019/commits/afd24cfb1715da85219874a09889081d6ba58c6f#r398541333

File: CODINGSTYLE.md
Comment (tomereli) [reply CODINGSTYLE.md:43]: Do you remember why we chose this?

File: CODINGSTYLE.md
Comment (tomereli) [inline CODINGSTYLE.md:34]: This is confusing, why did you call it `x_foo` and `y_foo` and not just `x` and `y`?

**Review by tomereli — COMMENTED** [review]
Same like Arnout said - all this has to be voted first on #prplmesh-democracy

File: CODINGSTYLE.md
Comment (morantr) [reply CODINGSTYLE.md:34]: To emphasize it is a lower snake case.

**Review by tomereli — APPROVED** [review]

**Review by arnout — APPROVED** [review]
Results from #prplmesh-democracy poll:

* Coding convention for classes: 9 CamelCase , 1 snake_case
* Naming convention for structs: 6 sPrefixedCamelCase , 1 snake_case, 1 CamelCase

Conclusion: this PR reflects the coding style we prefer.

**Pull request merged**
